image/image.go

```diff
@@ -50,17 +50,31 @@ var validRefMediaTypes = []string{
 }
 
 func validate(w walker, refs []string, out *log.Logger) error {
-	for _, r := range refs {
-		ref, err := findDescriptor(w, r)
-		if err != nil {
-			return err
+	ds, err := findDescriptors(w)
+	if err != nil {
+		return err
+	}
+	// TODO(runcom): this is O(N) but it could be O(1) if refs have been
+	// map[string]bool coming from the cli package.
+	// However, I don't think refs could be large enough to impact performance.
+	inRefs := func(ref string) bool {
+		for _, r := range refs {
+			if r == ref {
+				return true
+			}
 		}
-
-		if err = ref.validate(w, validRefMediaTypes); err != nil {
+		return false
+	}
+	hasRefs := len(refs) != 0
+	for ref, d := range ds {
+		if hasRefs && !inRefs(ref) {
```

Comment:
it's equivalent right? you still do an O(N) iterating over refs :(

I guess I'll change as you did if you really wish to